common/k8s/src/middleware/runner.rs

```diff
@@ -0,0 +1,111 @@
+use crate::{
+    create_k8s_client_default_from_env,
+    middleware::{K8sMetadata, StoreSwapIntent},
+};
+use futures::StreamExt;
+use hyper::http::header::HeaderValue;
+use middleware::Executor;
+use std::sync::Arc;
+use tokio::{
+    task::JoinHandle,
+    time::{sleep, Duration},
+};
+use tracing::{error, trace, warn};
+
+pub static K8S_WATCHER_ROTATION_DELAY: Duration = Duration::from_secs(3);
+
+enum State<'a> {
+    Init,
+    Creating(bool),
+    Kicking(bool, kube::Client),
+    Swapping(bool, JoinHandle<()>, StoreSwapIntent<'a>),
+    Running(JoinHandle<()>),
+}
+
+async fn step_trampoline<'a>(
+    state: State<'a>,
+    user_agent: &'a HeaderValue,
+    node_name: &'a Option<String>,
+    middleware: &'a K8sMetadata,
+) -> State<'a> {
+    match state {
+        State::Init => State::Creating(true),
+        State::Creating(first) => {
+            let middleware_k8s_client = create_k8s_client_default_from_env(user_agent.clone());
+            if let Err(e) = middleware_k8s_client {
+                let message = format!("unable to create client for k8s metadata watcher: {:?}", e);
+                error!("{}", message);
+                panic!("{}", message);
+            }
+            State::Kicking(first, middleware_k8s_client.unwrap())
```

Comment:
middleware_k8s_client always Some()? Err or Option?
may be match...

Reply:
The error case is handled in the lines before (it'll panic), so the unwrap should always have a `Some`. I simplified the code to process the error at the same time it unwraps it.